File: python/agents/hint_agent.py
"""
Hint Agent（提示Agent）-- 分级提示策略。

核心职责：
1. 收到提示请求后，根据学生尝试次数和mastery决定提示级别
2. 三级提示：暗示(Metacognitive) → 引导(Scaffolding) → 直接答案(Targeted)
3. 目标：85%的情况使用暗示或引导，仅15%给直接答案

面试要点：
- 分级提示的教育学依据：Vygotsky的最近发展区(ZPD)理论
- 为什么不直接给答案？研究表明引导式学习的记忆保留率是被动学习的3倍
- 提示级别如何决定？综合考虑尝试次数、mastery、时间消耗
"""

import logging

from .base_agent import BaseAgent
from core.event_bus import Event, EventType

logger = logging.getLogger(__name__)


class HintLevel:
    METACOGNITIVE = 1  # 暗示：引导学生反思思路
    SCAFFOLDING = 2  # 引导：给出关键步骤提示
    TARGETED = 3  # 直接：给出答案或具体解法


HINT_TEMPLATES = {
    HintLevel.METACOGNITIVE: {
        "description": "元认知暗示 - 引导学生反思自己的思考过程",
        "templates": [
            "想一想，这道题的题目里有哪些关键信息？你有没有全部用上？",
            "你能把这道题和之前学过的哪个类型联系起来吗？",
            "试试从结果反推，如果答案是XX，那过程应该是什么样的？",
            "画个图或者列个表，把已知条件整理一下，也许能看出规律。",
            "这道题有没有更简单的特殊情况？先从特殊情况入手试试。",
        ],
    },
    HintLevel.SCAFFOLDING: {
        "description": "脚手架引导 - 给出关键步骤但不给答案",
        "templates": [
            "这道题的关键是用到{concept}。你记得{concept}的公式/定义吗？",
            "第一步：先{step1}。第二步：然后{step2}。你试试看能不能完成。",
            "提示：答案和{hint_value}有关。试着从这个方向思考。",
            "这道题和{similar_problem}很像。回忆一下那道题是怎么做的？",
        ],
    },
    HintLevel.TARGETED: {
        "description": "直接提示 - 给出具体解法（仅在多次尝试后使用）",
        "templates": [
            "好的，这道题的解法是：{solution_steps}。\n关键点是{key_point}。\n建议你自己重新做一遍，确保理解了。",
        ],
    },
}


class HintAgent(BaseAgent):
    """提示Agent：实现分级提示策略。"""

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._hint_history: dict[str, dict[str, int]] = {}

    @property
    def subscribed_events(self) -> list[EventType]:
        return [EventType.HINT_NEEDED]

    async def handle_event(self, event: Event) -> None:
        if event.type == EventType.HINT_NEEDED:
            await self._provide_hint(event)

    def _determine_hint_level(
        self, learner_id: str, knowledge_id: str, mastery: float, attempts: int
    ) -> int:
        """
        决定提示级别的策略。

        规则：
        - 第1-2次尝试 → Level 1（元认知暗示）
        - 第3-4次尝试 → Level 2（脚手架引导）
        - 第5次及以上 → Level 3（直接提示）
        - mastery极低(< 0.15) 且 attempts >= 3 → 提前升级到 Level 3

        目标：85%使用Level 1-2，15%使用Level 3
        """
        key = f"{learner_id}:{knowledge_id}"
        hint_count = self._hint_history.get(key, 0)

        if mastery < 0.15 and attempts >= 3:
            return HintLevel.TARGETED

        if hint_count <= 1:
            return HintLevel.METACOGNITIVE
        elif hint_count <= 3:
            return HintLevel.SCAFFOLDING
        else:
            return HintLevel.TARGETED

    async def _provide_hint(self, event: Event) -> None:
        """生成并发送分级提示。"""
        learner_id = event.learner_id
        knowledge_id = event.data.get("knowledge_id", "")
        mastery = event.data.get("mastery", 0.0)
        attempts = event.data.get("attempts", 1)

        level = self._determine_hint_level(learner_id, knowledge_id, mastery, attempts)

        key = f"{learner_id}:{knowledge_id}"
        self._hint_history[key] = self._hint_history.get(key, 0) + 1

        hint_text = self._generate_hint(knowledge_id, level)

        level_names = {
            HintLevel.METACOGNITIVE: "metacognitive",
            HintLevel.SCAFFOLDING: "scaffolding",
            HintLevel.TARGETED: "targeted",
        }

        logger.info(
            "[HintAgent] learner=%s, kp=%s, level=%s (%d/%d)",
            learner_id,
            knowledge_id,
            level_names[level],
            self._hint_history[key],
            attempts,
        )

        await self.emit(
            EventType.HINT_RESPONSE,
            learner_id,
            {
                "knowledge_id": knowledge_id,
                "hint_level": level,
                "hint_level_name": level_names[level],
                "hint_text": hint_text,
                "hint_count": self._hint_history[key],
                "description": HINT_TEMPLATES[level]["description"],
            },
        )

    def _generate_hint(self, knowledge_id: str, level: int) -> str:
        """
        生成提示文本。

        生产环境中会调用LLM根据具体题目生成，
        这里用模板演示分级提示的逻辑。
        """
        templates = HINT_TEMPLATES[level]["templates"]
        template = templates[0]

        if level == HintLevel.METACOGNITIVE:
            return f"💡 关于「{knowledge_id}」的提示：\n{template}"
        elif level == HintLevel.SCAFFOLDING:
            return (
                f"📝 关于「{knowledge_id}」的引导提示：\n"
                f"这道题的关键是理解{knowledge_id}的核心概念。\n"
                f"试着回忆一下相关的公式或定义，然后一步步来。"
            )
        else:
            return (
                f"📖 关于「{knowledge_id}」的详细解答：\n"
                f"让我来帮你梳理一下解题思路。\n"
                f"这次我会给你更详细的指导，但请你一定要自己重做一遍。\n"
                f"【解题步骤】...\n"
                f"建议复习完后再做2道同类型的练习巩固。"
            )

File: java/src/main/java/com/edu/agent/agent/HintAgent.java
package com.edu.agent.agent;

import com.edu.agent.core.AgentEvent;
import com.edu.agent.core.EventType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.context.event.EventListener;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

/**
 * Hint Agent -- 三级提示策略。
 * Level 1: 元认知暗示  Level 2: 脚手架引导  Level 3: 直接答案
 */
@Component
public class HintAgent {

    private static final Logger log = LoggerFactory.getLogger(HintAgent.class);
    private final ApplicationEventPublisher publisher;
    private final Map<String, Integer> hintHistory = new ConcurrentHashMap<>();

    public HintAgent(ApplicationEventPublisher publisher) {
        this.publisher = publisher;
    }

    @Async
    @EventListener(condition = "#event.type == T(com.edu.agent.core.EventType).HINT_NEEDED")
    public void handleHintNeeded(AgentEvent event) {
        String learnerId = event.getLearnerId();
        String knowledgeId = (String) event.getData().getOrDefault("knowledge_id", "");
        double mastery = ((Number) event.getData().getOrDefault("mastery", 0.0)).doubleValue();
        int attempts = ((Number) event.getData().getOrDefault("attempts", 1)).intValue();

        int level = determineLevel(learnerId, knowledgeId, mastery, attempts);
        String key = learnerId + ":" + knowledgeId;
        hintHistory.merge(key, 1, Integer::sum);

        String hintText = generateHint(knowledgeId, level);
        String levelName = switch (level) {
            case 1 -> "metacognitive";
            case 2 -> "scaffolding";
            case 3 -> "targeted";
            default -> "unknown";
        };

        log.info("[Hint] learner={}, kp={}, level={}", learnerId, knowledgeId, levelName);

        Map<String, Object> data = new HashMap<>();
        data.put("knowledge_id", knowledgeId);
        data.put("hint_level", level);
        data.put("hint_level_name", levelName);
        data.put("hint_text", hintText);
        publisher.publishEvent(new AgentEvent(this, EventType.HINT_RESPONSE,
                "HintAgent", learnerId, data));
    }

    private int determineLevel(String learnerId, String knowledgeId, double mastery, int attempts) {
        String key = learnerId + ":" + knowledgeId;
        int hintCount = hintHistory.getOrDefault(key, 0);

        if (mastery < 0.15 && attempts >= 3) return 3;
        if (hintCount <= 1) return 1;
        if (hintCount <= 3) return 2;
        return 3;
    }

    private String generateHint(String knowledgeId, int level) {
        return switch (level) {
            case 1 -> String.format("💡 关于「%s」：想一想，题目里有哪些关键信息你还没用上？", knowledgeId);
            case 2 -> String.format("📝 关于「%s」：试着回忆相关的公式或定义，然后一步步来。", knowledgeId);
            case 3 -> String.format("📖 关于「%s」：让我帮你梳理解题思路，但请你一定要自己重做一遍。", knowledgeId);
            default -> "继续思考...";
        };
    }
}
